src/ui/public/courier/fetch/__tests__/call_client.js

```diff
@@ -0,0 +1,270 @@
+      const callingClient = callClient(searchRequests);
+
+      callingClient.then(results => {
+        expect(results).to.eql([1]);
+        done();
+      }).catch(error => done(error));
+    });
+  });
+
+  describe('implementation', () => {
+    it('calls es.msearch() once, regardless of number of searchRequests', done => {
+      expect(fakeSearch.callCount).to.be(0);
+
+      searchRequests = [ createSearchRequest(), createSearchRequest(), createSearchRequest() ];
+      const callingClient = callClient(searchRequests);
+
+      callingClient.then(() => {
+        expect(fakeSearch.callCount).to.be(1);
+        done();
+      }).catch(error => done(error));
+    });
+
+    it('calls searchRequest.whenAborted() as part of setup', async () => {
+      const whenAbortedSpy = sinon.spy();
+      const searchRequest = createSearchRequest();
+      searchRequest.whenAborted = whenAbortedSpy;
+      searchRequests = [ searchRequest ];
+      await callClient(searchRequests);
+      expect(whenAbortedSpy.callCount).to.be(1);
+    });
+
+    it(`calls searchRequest.handleFailure() with the ES error that's thrown`, async () => {
+      esShouldError = true;
+      const searchRequest = createSearchRequest(1);
+
+      const handleFailureSpy = sinon.spy();
+      searchRequest.handleFailure = handleFailureSpy;
+
+      searchRequests = [ searchRequest ];
+      await callClient(searchRequests);
+      sinon.assert.calledWith(handleFailureSpy, 'fake es error');
+    });
+  });
+
+  describe('aborting at different points in the request lifecycle:', () => {
+    it(`when searchSource's _flatten method throws an error resolves with an ABORTED response`, done => {
+      const searchRequest = createSearchRequest(1, {
+        source: {
+          _flatten: () => { throw new Error(); },
+        },
+      });
+
+      searchRequests = [ searchRequest ];
+      const callingClient = callClient(searchRequests);
+
+      callingClient.then(results => {
+        // The result is ABORTED because it was never included in the body sent to es.msearch().
+        expect(results).to.eql([ ABORTED ]);
+        done();
+      }).catch(error => done(error));
+    });
+
+    it('while the search body is being formed resolves with an ABORTED response', done => {
+      const searchRequest = createSearchRequest(1, {
+        source: {
+          _flatten: () => {
+            return new Promise(resolve => {
+              setTimeout(() => {
+                resolve({});
+              }, 100);
+            });
+          },
+          requestIsStopped: () => {},
+        },
+      });
+
+      searchRequests = [ searchRequest ];
+      const callingClient = callClient(searchRequests);
+
+      // Abort the request while the search body is being formed.
+      setTimeout(() => {
+        searchRequest.abort();
+      }, 20);
+
+      callingClient.then(results => {
+        expect(results).to.eql([ ABORTED ]);
+        done();
```

Comment:
Why use the callback interface when you can just return the promise?

```js
return callingClient.then(results => {
  expect(results).to.eql([ ABORTED ]);
})
```

Makes me more confident that errors will flow correctly

Reply:
Cool, I didn't know you can do that.